source/mir/combinatorics.d

```diff
@@ -0,0 +1,734 @@
+    }
+    return CartesianPower(r, repeat);
+}
+
+///
+pure nothrow @safe unittest
+{
+    import std.algorithm: equal;
+    import std.range: iota;
+    assert(iota(2).cartesianPower.equal!equal([[0], [1]]));
+    assert(iota(2).cartesianPower(2).equal!equal([[0, 0], [0, 1], [1, 0], [1, 1]]));
+
+    auto three_nums_two_bins = [[0, 0], [0, 1], [0, 2], [1, 0], [1, 1], [1, 2], [2, 0], [2, 1], [2, 2]];
+    assert(iota(3).cartesianPower(2).equal!equal(three_nums_two_bins));
+
+    assert("AB"d.cartesianPower(2).equal!equal(["AA"d, "AB"d, "BA"d, "BB"d]));
+}
+
+pure nothrow @safe unittest
+{
+    import std.algorithm: equal, map;
+    import std.array: array;
+    import std.range: iota, dropOne;
+
+    assert(iota(0).cartesianPower.length == 0);
+    assert("AB"d.cartesianPower(3).equal!equal(["AAA"d, "AAB"d, "ABA"d, "ABB"d, "BAA"d, "BAB"d, "BBA"d, "BBB"d]));
+    auto expected = ["AA"d, "AB"d, "AC"d, "AD"d,
+                     "BA"d, "BB"d, "BC"d, "BD"d,
+                     "CA"d, "CB"d, "CC"d, "CD"d,
+                     "DA"d, "DB"d, "DC"d, "DD"d];
+    assert("ABCD"d.cartesianPower(2).equal!equal(expected));
+    // verify with array too
+    assert("ABCD"d.cartesianPower(2).map!array.array == expected);
+
+    assert(iota(2).cartesianPower.front.equal([0]));
+
+    // is copyable?
+    auto a = iota(2).cartesianPower;
+    assert(a.front.equal([0]));
+    assert(a.save.dropOne.front.equal([1]));
+    assert(a.front.equal([0]));
+
+    // test length shrinking
+    auto d = iota(2).cartesianPower;
+    assert(d.length == 2);
+    d.popFront;
+    assert(d.length == 1);
+}
+
+/**
+Lazily computes all k-combinations of $(D r).
+Imagine this as the $(LREF cartesianPower) filtered for only strictly ordered items.
+
+While generating a new combination is in `O(k)`,
+the number of combinations is `binomial(|r|, k)`.
+
+For example $(D"AB".combinations(2).array) returns $(D["AB"]).
+
+Params:
+    r = $(REF_ALTTEXT Random access range, isRandomAccessRange, std, range, primitives)
+    k = number of combinations
+
+Returns:
+    Forward range which yields the k-combinations items
+*/
+auto combinations(Range)(Range r, size_t k = 1)
+if (isRandomAccessRange!Range && hasLength!Range)
+in
+{
+    assert(k >= 1, "Invalid number of combinations");
+}
+body
+{
+    import std.range: Indexed;
+    import std.range: iota;
+    static struct Combination
+    {
+    private:
+        Range _r;
+        size_t[] _state;
+        size_t _max_states, _pos;
+
+    public:
+        this(Range r, size_t repeat)
+        {
+            import std.array: array;
+
+            _r = r;
+
+            // set initial state and calculate max possibilities
+            if (r.length > 0)
+            {
+                _max_states = binomial(r.length, repeat);
+                _state = new size_t[](repeat);
+                // skip first duplicate
+                if (r.length > 1 && repeat > 1)
+                {
+                    _state = iota(repeat).array;
```

Comment:
This is not good. First you allocate an array with `new`. Than allocate another with `.array`. Please check another such cases.